Assets/Scripts/UI/AmmoUI.cs

```diff
@@ -0,0 +1,102 @@
+using System;
+using System.Collections.Generic;
+using TMPro;
+using Unity.VisualScripting;
+using UnityEngine;
+using UnityEngine.UI;
+
+public class AmmoUI : MonoBehaviour
+{
+    [SerializeField] private TextMeshProUGUI magazineLeftText;
+    [SerializeField] private HorizontalLayoutGroup bulletIconsLayoutGroup;
+
+    private readonly List<Image> bulletIcons = new List<Image>();
+    
+    // reload animation stuff
+    private float reloadTimer = 0f;
+    [NonSerialized] public bool isReloading = false;
+    [NonSerialized] public float reloadTime; // Time it takes to reload the gun
+
+    public int MagazineLeft
+    {
+        set
+        {
+            magazineLeftText.text = value.ToString();
+        }
+    }
+
+	public delegate void OnReloadFinish();
+	public static event OnReloadFinish OnReloadFinishEvent;
+
+    private void Update()
+    {
+        if (isReloading)
+        {
+            reloadTimer += Time.deltaTime;
+
+            float progress = reloadTimer / reloadTime;
+
+            // Update the reload bar fill amount
+            foreach (Image image in bulletIcons)
+            {
+                // only trigger animations on icons that are not filled
+                if (image.fillAmount != 1)
+                    image.fillAmount = progress;
+            }
+
+            if (reloadTimer >= reloadTime)
+            {
+                // If the gun has finished reloading
+                isReloading = false;
+                reloadTimer = 0;
+                OnReloadFinishEvent?.Invoke();
+            }
+        }
+    }
+
+    public void Rerender(Gun newGun)
+    {
+        foreach (Image icon in bulletIcons)
+            Destroy(icon.gameObject);
+
+        bulletIcons.Clear();
+
+        for (int i = 0; i < newGun.MaxShotPerMagazine; i++)
+        {
+            // create new game object and attach image component to it.
+            Image newIcon = new GameObject("BulletIcon" + i).AddComponent<Image>();
+            newIcon.sprite = newGun.GunSO.bulletPrefab.Icon;
+            newIcon.transform.parent = bulletIconsLayoutGroup.transform;
```

Comment:
Investigate Unity warning
> Parent of RectTransform is being set with parent property. Consider using the SetParent method instead, with the worldPositionStays argument set to false. This will retain local orientation and scale rather than world orientation and scale, which can prevent common UI scaling issues.

Reply:
Changed